d2/src/main/java/com/linkedin/d2/xds/LoadBalanceTaskPool/NewBalanceGetPropertiesTask.java

```diff
@@ -0,0 +1,85 @@
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.d2.balancer.LoadBalancerWithFacilities;
+import com.linkedin.d2.balancer.dualread.DualReadModeProvider;
+import com.linkedin.d2.balancer.dualread.DualReadStateManager;
+import com.linkedin.d2.balancer.properties.ClusterProperties;
+import com.linkedin.d2.balancer.properties.ServiceProperties;
+import com.linkedin.d2.balancer.properties.UriProperties;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class NewBalanceGetPropertiesTask implements Runnable {
+  private LoadBalancerWithFacilities _newLb;
+  private DualReadStateManager _dualReadStateManager;
+  private String serviceName;
+  private static final Logger LOG = LoggerFactory.getLogger(NewBalanceGetPropertiesTask.class);
+
+  public NewBalanceGetPropertiesTask(LoadBalancerWithFacilities _newLb, DualReadStateManager _dualReadStateManager,
+      String serviceName) {
+    this._newLb = _newLb;
+    this._dualReadStateManager = _dualReadStateManager;
+    this.serviceName = serviceName;
+  }
+
+  @Override
+  public void run() {
+    _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>() {
+      @Override
+      public void onError(Throwable e) {
+        LOG.error("Double read failure. Unable to read service properties from: " + serviceName, e);
+      }
+
+      @Override
+      public void onSuccess(ServiceProperties result) {
+        String clusterName = result.getClusterName();
+        _dualReadStateManager.updateCluster(clusterName, DualReadModeProvider.DualReadMode.DUAL_READ);
+        _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>() {
+          @Override
+          public void onError(Throwable e) {
+            LOG.error("Double read failure. Unable to read service properties from: " + serviceName, e);
+          }
+
+          @Override
+          public void onSuccess(ServiceProperties result) {
+            String clusterName = result.getClusterName();
+            _dualReadStateManager.updateCluster(clusterName, DualReadModeProvider.DualReadMode.DUAL_READ);
+            _newLb.getLoadBalancedClusterAndUriProperties(clusterName,
+                new Callback<Pair<ClusterProperties, UriProperties>>() {
+                  @Override
+                  public void onError(Throwable e) {
+                    LOG.error("Dual read failure. Unable to read cluster properties from: " + clusterName, e);
+                  }
+
+                  @Override
+                  public void onSuccess(Pair<ClusterProperties, UriProperties> result) {
+                    //TODO change back to debug
+                    LOG.info("Dual read is successful. Get cluster and uri properties: " + result);
+                  }
+                });
+          }
+        });
+      }
+    });
```

Comment:
I'm definitely in favor of leaving this as a lambda inside `DualReadLoadBalancer`. If you leave this as a public class, it makes the API for this package unclear and confusing. This is an implementation detail that should not be exposed to users of this package. If you want, you can make it a function of `DualReadLoadBalancer` then invoke it like this:
```java
_newNewLoadBalancerTaskThreadPool.execute(() -> getLoadBalancedServiceProperties(serviceName));
...

private void getLoadBalancedServiceProperties(String serviceName) {
    _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>() {
      @Override
      public void onError(Throwable e) {
        LOG.error("Double read failure. Unable to read service properties from: " + serviceName, e);
      }

      @Override
      public void onSuccess(ServiceProperties result) {
        String clusterName = result.getClusterName();
        _dualReadStateManager.updateCluster(clusterName, DualReadModeProvider.DualReadMode.DUAL_READ);
        _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>() {
          @Override
          public void onError(Throwable e) {
            LOG.error("Double read failure. Unable to read service properties from: " + serviceName, e);
          }

          @Override
          public void onSuccess(ServiceProperties result) {
            String clusterName = result.getClusterName();
            _dualReadStateManager.updateCluster(clusterName, DualReadModeProvider.DualReadMode.DUAL_READ);
            _newLb.getLoadBalancedClusterAndUriProperties(clusterName,
                new Callback<Pair<ClusterProperties, UriProperties>>() {
                  @Override
                  public void onError(Throwable e) {
                    LOG.error("Dual read failure. Unable to read cluster properties from: " + clusterName, e);
                  }

                  @Override
                  public void onSuccess(Pair<ClusterProperties, UriProperties> result) {
                    //TODO change back to debug
                    LOG.info("Dual read is successful. Get cluster and uri properties: " + result);
                  }
                });
          }
        });
}
```

Reply:
+1, those are good arguments.

Reply:
Updated.